tests/ClientServer.h

```diff
@@ -0,0 +1,160 @@
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#pragma once
+
+#include <brayns/Brayns.h>
+#include <brayns/parameters/ParametersManager.h>
+
+#include <boost/test/unit_test.hpp>
+
+#include <rockets/jsonrpc/client.h>
+#include <rockets/ws/client.h>
+
+#include <future>
+
+template <typename T>
+bool is_ready(const std::future<T>& f)
+{
+    return f.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
+}
+
+class ClientServer
+{
+public:
+    static ClientServer& instance()
+    {
+        if (!_instance)
+            throw std::runtime_error("Could not initialize client/server");
+        return *_instance;
+    }
+
+    ClientServer(
+        std::vector<const char*> additionalArgv = std::vector<const char*>())
+    {
+        auto& testSuite = boost::unit_test::framework::master_test_suite();
+        const char* app = testSuite.argv[0];
+        std::vector<const char*> argv{app, "--http-server", ":0"};
+        for (const auto& arg : additionalArgv)
+            argv.push_back(arg);
+        const int argc = argv.size();
+        brayns.reset(new brayns::Brayns(argc, argv.data()));
+        brayns->loadPlugins();
+        brayns->getParametersManager()
+            .getApplicationParameters()
+            .setImageStreamFPS(0);
+        brayns->render();
+
+        const auto uri = brayns->getParametersManager()
+                             .getApplicationParameters()
+                             .getHttpServerURI();
+
+        auto connectFuture = wsClient.connect("ws://" + uri, "rockets");
+        while (!is_ready(connectFuture))
+        {
+            wsClient.process(5);
+            brayns->render();
+        }
+        connectFuture.get();
+        _instance = this;
+    }
+
+    template <typename Params, typename RetVal>
+    RetVal makeRequest(const std::string& method, const Params& params)
+    {
+        auto responseFuture = client.request<Params, RetVal>(method, params);
+        while (!is_ready(responseFuture))
+        {
+            wsClient.process(0);
+            brayns->render();
+        }
+
+        return responseFuture.get();
+    }
+
+    template <typename RetVal>
+    RetVal makeRequest(const std::string& method)
+    {
+        auto responseFuture = client.request<RetVal>(method);
+        while (!is_ready(responseFuture))
+        {
+            wsClient.process(0);
+            brayns->render();
+        }
+
+        return responseFuture.get();
+    }
+
+    template <typename Params>
+    void makeNotification(const std::string& method, const Params& params)
+    {
+        client.notify<Params>(method, params);
+
+        wsClient.process(5);
+        for (size_t i = 0; i < 10; ++i)
```

Comment:
Magic numbers 5 and 10? Can you define constants for that with an explicit name?

Reply:
no :)